tests/scripted/check-namespace

```diff
@@ -0,0 +1,199 @@
+#!/usr/bin/python3
+
+# This script checks the code for namespace related issues
+# Note: This script is not perfect. It can not parse all LMMS files,
+#       and does not contain a complete C++ parser. If you encounter
+#       difficulties with this tests, it could be a fault of your
+#       changes, but it could also be a fault of this script.
+
+import re
+import subprocess
+from pathlib import Path
+
+# global variables
+
+errors = 0
+
+
+# functions
+
+def caption(my_str):
+    print(f'\n# {my_str}\n')
+
+
+def error(where, my_str):
+    global errors
+    errors += 1
+    print(f'Error: {where}: {my_str}')
+
+
+# if a string makes a classname, we need to check if that class is in the source
+# however, some of these strings name classes that are not from LMMS, so these can be ignored for such checks:
+def is_our_class(classname: str) -> int:
+    return classname[0] != 'Q'  # Qt classes
+
+
+# prepare some variables
+
+if not Path('.gitmodules').is_file():
+    print('You need to call this script from the LMMS top directory')
+    exit(1)
+
+result = subprocess.run(['git', 'ls-files', '*.[ch]', '*.[ch]pp', ':!tests/*'],
+                        capture_output=True, text=True, check=True)
+exclude_files = {
+    # too complex
+    'include/AudioPortAudio.h',
+    'src/core/audio/AudioPortAudio.cpp',
+    'include/RemotePluginBase.h',
+    'plugins/VstBase/RemoteVstPlugin.cpp',
+    'plugins/ReverbSC/ReverbSC.h',
+    # not ours:
+    'include/ladspa.h'
+    # f.startswith('plugins/LadspaEffect')
+    # f.__contains__('portsmf')
+}
+known_no_namespace_lmms = {
+    # main.cpp
+    'src/core/main.cpp',
+    # nothing to set under a namespace
+    'include/debug.h',
+    'include/versioninfo.h',
+    'plugins/CarlaBase/CarlaConfig/config.h',
+    'plugins/CarlaBase/DummyCarla.cpp',
+    # unclear why it has no namespace
+    'plugins/ZynAddSubFx/RemoteZynAddSubFx.cpp',
+    # not our code
+    'include/aeffectx.h',
+    'include/fenv.h',
+    'plugins/Sf2Player/fluidsynthshims.h',
+    'plugins/MidiExport/MidiFile.hpp',
+    'plugins/ReverbSC/base.c',
+    'plugins/ReverbSC/base.h',
+    'plugins/ReverbSC/dcblock.c',
+    'plugins/ReverbSC/dcblock.h',
+    'plugins/ReverbSC/revsc.c',
+    'plugins/ReverbSC/revsc.h'
+}
+
+files = [Path(f) for f in result.stdout.splitlines() if
+            f not in exclude_files
+            and not f.startswith('plugins/LadspaEffect/calf')
+            and not f.startswith('plugins/LadspaEffect/caps')
+            and not f.startswith('plugins/LadspaEffect/cmt')
+            and not f.startswith('plugins/LadspaEffect/swh')
+            and not f.startswith('plugins/LadspaEffect/tap')
+            and not f.__contains__('portsmf')
+        ]
+
+namespaces = re.compile(r'^\s*namespace\s*\S*', re.MULTILINE)
+no_include_after_namespace = re.compile(
+    r'^(\s*#\s*(?:include|if|el|endif)[^\n]*)*(namespace [\w:]+ [{]|#(?:ifn?def(?: [\w_]+)|if|endif|el)[a-z]*(?: // ?[\w_]+)?|[}](?: // namespace [\w:]+)?|\n)*$')
```

Comment:
This is a very complex regex, so I'm not sure I'm getting it right, but is the part at the beginning checking for any of `#include`, `#if`, `#elif` and `#endif` inside a namespace scope? If so, it's probably going to flag a lot of files that use conditional compilation in the middle of class and/or function implementations.

Reply:
No, the first part allows all kind of macros (`#...`), including `#include`. And after that, only `namespace` and `#if...` (as well as closing }) can occur, but no `#include` anymore. So if this does not match, we have `#include` after the first occurence of `namespace` (which is why the variable is named like that).

Reply:
@JohannesLorenz Can you provide some examples where the regex matches(or not), and test this regex with the examples?

Reply:
@PhysSong I added the only test case I thought could make sense, please check [this link](https://github.com/JohannesLorenz/lmms/blob/f1f93b10fa0075d83f706c836b993c66568e17e0/tests/scripted/verify#L194-L199).

Reply:
Looking into the regex, I found some potential issues:
- `namespace [\w:]+ `: What if someone writes in wrong formatting such as `namespace  lmms{`?
- ` // `: Similar, like `}//namespace lmms`.
- The regex should match multiple lines, but you haven't specified `re.MULTILINE` while using `^` and `$`.

Not sure if there are more, because I haven't tested this with many test cases yet.